### ISSUE TITLE
Add resolveNestedSelectors:true option to selector-class-pattern

### ISSUE THREAD

**Issue by onigoetz** [stylelint/stylelint#835]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Selector class pattern doesn't match on nested elements with "&"

> Which rule, if any, is this issue related to?

`selector-class-pattern`

> What CSS is needed to reproduce this issue?

```css
.MyOtherComponent {
    &__invalid-element {
        color: orange;
    }
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-class-pattern": /^(_)?(?:(?:u|t|(?:i|ha)?s)+-[a-z]+[a-zA-Z0-9]*|[A-Z][a-zA-Z0-9]*(__[a-z][a-zA-Z0-9]+)*(--[a-z][a-zA-Z0-9]+)?)$/
  }
}
```

> Which version of stylelint are you using?

`4.4.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with the config shown before

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to nesting

> What did you expect to happen?

The following warnings to be flagged:

```shell
test.css
2:5    Expected class selector ".MyOtherComponent__invalid-element" to match specified pattern (selector-class-pattern)
```

> What actually happened ?

No warnings were flagged.

This is also related to #834 as the sub-element is recognized as a `type` and not a `class`

---
Edit : Inverted the order of "What did you expect to happen?" and "What actually happened", was wrong before

**Comment by jeddy3** [comment]
Are you running `stylelint` before or after your processors?

The warning above implies you're running it after as `".MyOtherComponent__invalid-element"` is shown in the warning and we don't resolve the nesting selector in this rule.

Your warning in https://github.com/stylelint/stylelint/issues/834 implies the opposite though i.e. your running `stylelint` before any processors as it's tripping up on the nesting selector being used as a prefix.

Confirmation of whether your running `stylelint` before or after your processor would help us out with this particular issue :)

**Comment by onigoetz** [comment]
> After writing this comment I discovered that I put in the wrong order the " what did I expect " and "what happened" ... I will edit the issue


I'm running it completely separately.

Actually, my linter configuration is made to run at two times : on "compilation" with gulp, or on mercurial push with a hook.

I both cases, I can't do it after, because some rules don't react the same if I'm working on compiled code (I use `autoprefixer`, `cssnano`, `cssnext` ) and I have formatting rules, so for example `cssnano` can't really be run before the formatting rules ...

Also, I use the `statement-max-nesting-depth` plugin, which also needs to run before compilation.

**Comment by jeddy3** [comment]
> After writing this comment I discovered that I put in the wrong order the " what did I expect " and "what happened" ... I will edit the issue

Oooo, that explains it :) I now understand what you're asking for.

@davidtheclark This is a request to resolve the nesting selector (`&`) before testing the pattern. Up until recently none of the selector rules resolved `&`, but now we have three rules that do it:

- `no-duplicate-selectors`
- `selector-no-descending-specificity`
- `selector-max-specificity`

So `&` is currently treated inconsistently. I'm a little unsure what's best here and I can think of a few options:

- Prior to releasing `4.5`, we make resolving `&` *an option* for the three rules above i.e. the default behaviour isn't to resolve the nesting selector. We can then add an option for resolving `&` to each of the older selector rules on a per-request basis (starting with `selector-class-pattern`).
- We make resolving `&` the default *and only* behaviour across the older selector rules (breaking change - we've making the old behaviour unavailable).
- We make resolving `&` the default behaviour across the rules, but then add *an option* to both the older and three new rules to **not** resolve `&` (breaking change, but the old behaviour is still available behind an option).

I think my gut is saying the 1st option, i.e. explicitly opting in to resolving the selector, as, I believe, that behaviour is more consistent with how the linter works across the board. We tend to just look for simple patterns in the source. Whereas resolving the nesting selector is doing something more and involves matching patterns against *what the source will become*. What do you think?

**Comment by onigoetz** [comment]
Yeah, sorry for that I was in heavy copy-pasting mode, and that one slipped through :p 

Well, clearly with the context of a history and users relying on existing behaviour, the 1st options would be the one to go with.

**Comment by davidtheclark** [comment]
Sure. I'd say `resolveNestedSelectors` would be a good option name?

**Comment by onigoetz** [comment]
yes, I agree that's a good name

**Comment by jeddy3** [comment]
> I'd say `resolveNestedSelectors` would be a good option name?

Sgtm.

The 1st option above suggested changing the following three rules so that they don't resolve by default, and instead only resolve when a proposed `resolveNestedSelectors` option is set:

- `no-duplicate-selectors`
- `selector-no-descending-specificity`
- `selector-max-specificity`

I've had a change of heart about this as not resolving `&` doesn't really make sense in these three rules. For example, we can't calculate a selector specificity if it contains an `&`. Likewise, we can't accurately match duplicates without first resolving `&`. 

So, I think it's OK that some rules resolve by default and some don't. We can make it clear (if it's not already so) in the docs for these three rules that they resolve `&` by default. How does that sound?

I'm setting this issue to "option" and "help wanted". This is for the task of just adding a `resolveNestedSelectors` option to `selector-class-pattern`. @onigoetz Feel free to jump on it if you'd like to contribute :)
